File: sdks/flutter/lib/cache.dart
class Cache<T> {
  final Map<String, _CacheEntry<T>> _cache = {};
  final Duration expireAfter;

  Cache({required this.expireAfter});

  void set(String key, T value) {
    _cache[key] = _CacheEntry(
      value: value,
      expiresAt: DateTime.now().add(expireAfter),
    );
  }

  T? get(String key) {
    final entry = _cache[key];
    if (entry == null) return null;

    if (DateTime.now().isAfter(entry.expiresAt)) {
      _cache.remove(key);
      return null;
    }

    return entry.value;
  }

  void invalidate(String key) {
    _cache.remove(key);
  }

  void invalidateAll() {
    _cache.clear();
  }

  int get size => _cache.length;
}

class _CacheEntry<T> {
  final T value;
  final DateTime expiresAt;

  _CacheEntry({required this.value, required this.expiresAt});
}

File: sdks/typescript/src/Cache.ts
export class Cache<T> {
  private cache = new Map<string, { value: T; expiresAt: number }>();
  private readonly expireAfterMs: number;

  constructor(expireAfterMs: number) {
    this.expireAfterMs = expireAfterMs;
  }

  set(key: string, value: T): void {
    this.cache.set(key, {
      value,
      expiresAt: Date.now() + this.expireAfterMs,
    });
  }

  // Allow type override when getting values for better type safety
  get<U = T>(key: string): U | undefined {
    const entry = this.cache.get(key);
    if (!entry) return undefined;

    if (Date.now() > entry.expiresAt) {
      this.cache.delete(key);
      return undefined;
    }

    return entry.value as unknown as U;
  }

  invalidate(key: string): void {
    this.cache.delete(key);
  }

  invalidateAll(): void {
    this.cache.clear();
  }

  get size(): number {
    return this.cache.size;
  }
}
